python/ray/train/data_parallel_trainer.py

```diff
@@ -396,66 +396,54 @@ def _report(self, training_iterator: TrainingIterator) -> None:
-
-                # Make sure that all workers uploaded to the same location.
-                assert all(
-                    checkpoint.path == tune_session.storage.checkpoint_fs_path
-                    for checkpoint in worker_checkpoints
-                ), (worker_checkpoints, tune_session.storage.checkpoint_fs_path)
-
-                checkpoint = (
-                    Checkpoint(
-                        filesystem=tune_session.storage.storage_filesystem,
-                        # NOTE: The checkpoint index has not been incremented yet
-                        # at this point, which is why `checkpoint_fs_path` points
-                        # to the most recent checkpoint.
-                        path=tune_session.storage.checkpoint_fs_path,
-                    )
-                    if at_least_one_reported_checkpoint
-                    else None
+            assert all(isinstance(result, _TrainingResult) for result in results)
+
+            tune_session = get_session()
+
+            # Check if any workers reported a checkpoint.
+            # If so, report a checkpoint pointing to the persisted location
+            # to Tune for book-keeping.
+            # NOTE: This removes the restriction for any individual worker
+            # (ex: global rank 0 worker) from needing to report a checkpoint.
+            # All workers reported a checkpoint to the same fs path, so there's
+            # no need to report multiple checkpoints to Tune.
+            worker_checkpoints = [
+                result.checkpoint for result in results if result.checkpoint is not None
+            ]
+            at_least_one_reported_checkpoint = len(worker_checkpoints) > 0
+
+            if at_least_one_reported_checkpoint:
+                # NOTE: This is where the coordinator AND workers increment their
+                # checkpoint index.
+                tune_session.storage._update_checkpoint_index(
+                    first_worker_result.metrics
                 )
 
-                metrics = first_worker_result.metrics.copy()
-                if at_least_one_reported_checkpoint:
-                    metrics[
-                        CHECKPOINT_DIR_NAME
-                    ] = tune_session.storage.checkpoint_dir_name
+            # Make sure that all workers uploaded to the same location.
+            assert all(
+                checkpoint.path == tune_session.storage.checkpoint_fs_path
+                for checkpoint in worker_checkpoints
+            )
 
-                tracked_training_result = _TrainingResult(
-                    checkpoint=checkpoint,
-                    metrics=metrics,
+            checkpoint = (
+                Checkpoint(
+                    filesystem=tune_session.storage.storage_filesystem,
+                    # NOTE: The checkpoint index has not been incremented yet
+                    # at this point, which is why `checkpoint_fs_path` points
+                    # to the most recent checkpoint.
```

Comment:
nit: This note is not as relevant now that the update happens right before uploading